kernel/byterun/coq_uint63_native.h

```diff
@@ -109,37 +110,56 @@ value uint63_mulc(value x, value y, value* h) {
 #define lt128(xh,xl,yh,yl) (uint63_lt(xh,yh) || (uint63_eq(xh,yh) && uint63_lt(xl,yl)))
 #define le128(xh,xl,yh,yl) (uint63_lt(xh,yh) || (uint63_eq(xh,yh) && uint63_leq(xl,yl)))
 
-value uint63_div21(value xh, value xl, value y, value* q) {
-  xh = (uint64_t)xh >> 1;
-  xl = ((uint64_t)xl >> 1) | ((uint64_t)xh << 63);
-  xh = (uint64_t)xh >> 1;
+#define maxuint63 ((uint64_t)0x7FFFFFFFFFFFFFFF)
+/* precondition: y <> 0 */
+/* outputs r and sets ql to q % 2^63 s.t. x = q * y + r, r < y */
+static value uint63_div21_aux(value xh, value xl, value y, value* ql) {
+  xh = uint63_of_value(xh);
+  xl = uint63_of_value(xl);
+  y = uint63_of_value(y);
   uint64_t maskh = 0;
   uint64_t maskl = 1;
   uint64_t dh = 0;
-  uint64_t dl = (uint64_t)y >> 1;
+  uint64_t dl = y;
   int cmp = 1;
-  while (dh >= 0 && cmp) {
+  /* int n = 0 */
+  /* loop invariant: mask = 2^n, d = mask * y, (2 * d <= x -> cmp), n >= 0, d < 2^(2*63) */
```

Comment:
I’m not convinced by the invariant « 2d ≤ x ⇒ cmp ».

Reply:
It holds before entering the loop: cmp is initialized to true

It remains true after each loop iteration: 2d ≤ x ⇒ d < x (since d = 2^n * y > 0 (precondition of the function)) and d < x = cmp, thus 2d ≤ x ⇒ cmp

When exiting the loop, we have:
* either dh >> (63 - 1) = 1, hence 2d >= 2^(2 * 63) > x (since x fits in 2 * 63 bits)
* or cmp = 0, meaning by contraposing the loop invariant that x < 2d

and r = x < 2d = 2^(n+1) <= y * 2^(n+1) enables to initialize the next loop invariant

I agree it looks a bit convoluted, feel free to offer anything better.

Reply:
Thanks for the clarification.